pkg/build/buildkit/k8s_autodiscovery.go

```diff
@@ -0,0 +1,338 @@
+		return "", err
+	}
+
+	// See more about App resource at: https://github.com/tsuru/tsuru/blob/main/provision/kubernetes/pkg/apis/tsuru/v1/types.go#L24
+	ns, found, err := unstructured.NestedString(tsuruApp.Object, "spec", "namespaceName")
+	if err != nil {
+		return "", err
+	}
+
+	if !found {
+		return "", fmt.Errorf("failed to fetch namespace in the App resource")
+	}
+
+	klog.V(4).Infof("App %s is running on namespace %s...", app, ns)
+
+	return ns, nil
+}
+
+func watchBuildKitPods(ctx context.Context, cs *kubernetes.Clientset, labelSelector, ns string, pods chan<- *corev1.Pod) error {
+	w, err := cs.CoreV1().Pods(ns).Watch(ctx, metav1.ListOptions{
+		LabelSelector: labelSelector,
+		Watch:         true,
+	})
+	if err != nil {
+		return err
+	}
+
+	defer w.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case e := <-w.ResultChan():
+			if e.Type != watch.Added && e.Type != watch.Modified {
+				continue
+			}
+
+			pod := e.Object.(*corev1.Pod)
+			if isPodReady(pod) {
+				pods <- pod
+			}
+
+			klog.V(4).Infof("Pod %s/%s is not ready yet", pod.Namespace, pod.Name)
+		}
+	}
+}
+
+func acquireLeaseForPod(ctx context.Context, cs *kubernetes.Clientset, ch chan<- *corev1.Pod, pod *corev1.Pod, opts KubernertesDiscoveryOptions) {
+	podname := os.Getenv("POD_NAME")
+	if podname == "" {
+		podname, _ = os.Hostname()
+	}
+
+	klog.V(4).Infof("Attempting to acquire the lease for pod %s/%s under holder name %q...", pod.Namespace, pod.Name, podname)
+
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock: &resourcelock.LeaseLock{
+			LeaseMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%s", strings.TrimRight(opts.LeasePrefix, "-"), pod.Name),
+				Namespace: pod.Namespace,
+			},
+			Client: cs.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: podname,
+			},
+		},
+		ReleaseOnCancel: true,
+		LeaseDuration:   5 * time.Second,
+		RenewDeadline:   2 * time.Second,
+		RetryPeriod:     500 * time.Millisecond,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(c context.Context) {
+				select {
+				case ch <- pod:
+					klog.V(4).Infof("Selected BuildKit pod: %s/%s", pod.Namespace, pod.Name)
+
+				case <-ctx.Done():
+					klog.V(4).Infof("Received context cancelation: %s/%s", pod.Namespace, pod.Name)
+				}
+			},
+			OnStoppedLeading: noopFunc,
+		},
+	})
+
+	klog.V(4).Infof("Shutting off the lease for %s/%s pod", pod.Namespace, pod.Name)
+}
+
+func setTsuruAppLabelOnBuildKitPod(ctx context.Context, cs *kubernetes.Clientset, pod, ns, app string) error {
+	patch, err := json.Marshal([]any{
+		map[string]any{
+			"op":    "replace",
+			"path":  fmt.Sprintf("/metadata/labels/%s", normalizeAppLabelForJSONPatch(TsuruAppNameLabelKey)),
+			"value": app,
+		},
+		map[string]any{
+			"op":    "replace",
+			"path":  fmt.Sprintf("/metadata/labels/%s", normalizeAppLabelForJSONPatch(TsuruIsBuildLabelKey)),
+			"value": strconv.FormatBool(true),
+		},
+	})
```

Comment:
@nettoclaudio we could also add a annotation to track the time that build started, this could be useful to scale to 0 when no build is running for a long time.

Reply:
Yes, we do. Do you want an annotation for the starting and ending build time, right?